internal/controller/nemoservice_controller.go

```diff
@@ -0,0 +1,586 @@
+	service NemoServiceSpec[T],
+	obj client.Object,
+) error {
+	logger := log.FromContext(ctx)
+
+	if service.Enabled == nil || !*service.Enabled {
+		return nil
+	}
+
+	// Set metadata
+	obj.SetName(service.Name)
+	obj.SetNamespace(nemoService.Namespace)
+
+	// Set service spec
+	err := SetObjectSpec(obj, service)
+	if err != nil {
+		logger.Error(err, "Failed to set object spec")
+		return err
+	}
+
+	// Set controller reference
+	if err := controllerutil.SetControllerReference(nemoService, obj, r.Scheme); err != nil {
+		return err
+	}
+
+	// Fetch current state
+	current := obj.DeepCopyObject().(client.Object)
+	err = r.Get(ctx, types.NamespacedName{Name: service.Name, Namespace: nemoService.Namespace}, current)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if errors.IsNotFound(err) {
+		logger.Info("Creating new service", "name", service.Name)
+		return r.Create(ctx, obj)
+	}
+
+	// Preserve resource version for update
+	obj.SetResourceVersion(current.GetResourceVersion())
+
+	logger.Info("Updating existing service", "name", service.Name)
+	return r.Update(ctx, obj)
+}
+
+// Helper function to extract service spec
+func getServiceSpec(nemoService *appsv1alpha1.NemoService, serviceType ServiceType) NemoServiceSpec[any] {
+	switch serviceType {
+	case Customizer:
+		return NemoServiceSpec[any]{nemoService.Spec.Customizer.Enabled, nemoService.Spec.Customizer.Name, nemoService.Spec.Customizer.Spec}
+	case Datastore:
+		return NemoServiceSpec[any]{nemoService.Spec.Datastore.Enabled, nemoService.Spec.Datastore.Name, nemoService.Spec.Datastore.Spec}
+	case Evaluator:
+		return NemoServiceSpec[any]{nemoService.Spec.Evaluator.Enabled, nemoService.Spec.Evaluator.Name, nemoService.Spec.Evaluator.Spec}
+	case Guardrails:
+		return NemoServiceSpec[any]{nemoService.Spec.Guardrails.Enabled, nemoService.Spec.Guardrails.Name, nemoService.Spec.Guardrails.Spec}
+	case Entitystore:
+		return NemoServiceSpec[any]{nemoService.Spec.Entitystore.Enabled, nemoService.Spec.Entitystore.Name, nemoService.Spec.Entitystore.Spec}
+	default:
+		return NemoServiceSpec[any]{nil, "", nil}
+	}
+}
+
+func (r *NemoServiceReconciler) cleanupDisabledNemoMicroservices(ctx context.Context, nemoService *appsv1alpha1.NemoService, enabledServices map[string]bool) error {
+	logger := log.FromContext(ctx)
+
+	var allErrors []error
+
+	// Define the list of Nemo microservice types to check for cleanup
+	serviceLists := map[ServiceType]client.ObjectList{
+		Customizer:  &appsv1alpha1.NemoCustomizerList{},
+		Datastore:   &appsv1alpha1.NemoDatastoreList{},
+		Evaluator:   &appsv1alpha1.NemoEvaluatorList{},
+		Guardrails:  &appsv1alpha1.NemoGuardrailList{},
+		Entitystore: &appsv1alpha1.NemoEntitystoreList{},
+	}
+
+	// Iterate over each Nemo microservice type and clean up disabled instances
+	for serviceType, serviceList := range serviceLists {
+		if err := r.List(ctx, serviceList, client.InNamespace(nemoService.Namespace)); err != nil {
```

Comment:
Instead of checking owner references later, we should filter here to only list objects owned by the parent CR (its a good practice to reduce memory footprint)
```suggestion
		if err := r.List(ctx, serviceList, client.InNamespace(nemoService.Namespace), client.MatchingLabels{<key>: <value>}); err != nil {
```
Instead of labels, we could also match owner references, but it would require setting up an index for it.